utils/wasm-instrument/src/tests.rs

```diff
@@ -0,0 +1,597 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use elements::Instruction::*;
+use gas_metering::ConstantCostRules;
+use parity_wasm::serialize;
+
+fn get_function_body(module: &elements::Module, index: usize) -> Option<&[elements::Instruction]> {
+    module
+        .code_section()
+        .and_then(|code_section| code_section.bodies().get(index))
+        .map(|func_body| func_body.code().elements())
+}
+
+fn prebuilt_simple_module() -> elements::Module {
+    builder::module()
+        .global()
+        .value_type()
+        .i32()
+        .build()
+        .function()
+        .signature()
+        .param()
+        .i32()
+        .build()
+        .body()
+        .build()
+        .build()
+        .function()
+        .signature()
+        .param()
+        .i32()
+        .build()
+        .body()
+        .with_instructions(elements::Instructions::new(vec![
+            Call(0),
+            If(elements::BlockType::NoResult),
+            Call(0),
+            Call(0),
+            Call(0),
+            Else,
+            Call(0),
+            Call(0),
+            End,
+            Call(0),
+            End,
+        ]))
+        .build()
+        .build()
+        .build()
+}
+
+#[test]
+fn simple_grow() {
+    let module = parse_wat(
+        r#"(module
+        (func (result i32)
+            global.get 0
+            memory.grow)
+        (global i32 (i32.const 42))
+        (memory 0 1)
+        )"#,
+    );
+
+    let injected_module = inject(module, &ConstantCostRules::new(1, 10_000), "env").unwrap();
+
+    // two new imports (indexes 0 & 1), the original func (i = 2), so
+    // gas charge will occupy the next index.
+    let gas_charge_index = 3;
+    let grow_index = 4;
+
+    assert_eq!(
+        get_function_body(&injected_module, 0).unwrap(),
+        &vec![
+            I32Const(2),
+            Call(gas_charge_index),
+            GetGlobal(0),
+            Call(grow_index),
+            End
+        ][..]
+    );
+    assert_eq!(
+        get_function_body(&injected_module, 2).unwrap(),
+        &vec![
+            GetLocal(0),
+            GetLocal(0),
+            I32Const(10_000),
+            I32Mul,
+            Call(gas_charge_index),
+            GrowMemory(0),
+            End,
+        ][..]
+    );
+
+    let binary = serialize(injected_module).expect("serialization failed");
+    wasmparser::validate(&binary).unwrap();
+}
+
+#[test]
+fn grow_no_gas_no_track() {
+    let module = parse_wat(
+        r"(module
+        (func (result i32)
+            global.get 0
+            memory.grow)
```

Comment:
Why so many tests with memory.grow ? this instruction is restricted in gear.

Reply:
These are the original tests from wasm-instrument

Reply:
Ok